feat: implement stablesats for galoy connector #2730

riccardobl opened the pull request:
### Describe the changes you have made in this PR
This is my submission for the galoy bounty https://github.com/GaloyMoney/galoy/discussions/3950  and feature request #2730 
The PR adds support for stable sats in the galoy connector.

### Link this PR to an issue [optional]
Implements #2730 

### Type of change


- `feat`: New feature (non-breaking change which adds functionality)


### Screenshots of the changes [optional]

![image](https://github.com/getAlby/lightning-browser-extension/assets/4943530/6717cae7-4535-4eec-ba49-99a49be06265)
![image](https://github.com/getAlby/lightning-browser-extension/assets/4943530/b048891f-326b-455e-a85e-25306cd38686)

### How has this been tested?
1. Connect a Blink account
2. When prompted select the `USD (Stablesats)` wallet
3. Send and receive stablesats to and from another lightning wallet


### Checklist

- [x] Self-review of changed code
- [x] Manual testing
- [ ] Added automated tests where applicable
- [ ] Update Docs & Guides
- For UI-related changes
- [ ] Darkmode
- [ ] Responsive layout

reneaaron commented:
@openoms May I ask you for a quick review? I am also planning to test / review it DURING THE NEXT FEW DAYS.

riccardobl commented:
@openoms To address the points as requested, this PR needs to change several things in the alby core, i've already prepared a draft commit that implements these changes https://github.com/getAlby/lightning-browser-extension/pull/3020/commits/92eeeec632921fe9ff4ace301c0c8fc6722398f8 ,  I'd like a feedback from the Alby team,  @reneaaron, regarding the overall concept of these changes, if we reach a consensus, I'll refine the commit to align with the code quality of the repository.

These are the screenshot after the last commit:

![image](https://github.com/getAlby/lightning-browser-extension/assets/4943530/5964a2b6-e33d-41b7-85e7-eb165d5f301e)
![image](https://github.com/getAlby/lightning-browser-extension/assets/4943530/2793593e-184d-4d76-88ac-d9221aee76a9)
![image](https://github.com/getAlby/lightning-browser-extension/assets/4943530/ca5e453c-bd2b-40c5-a496-d01c73c30ed5)

riccardobl commented:
I've cleaned and refactored the proposal with the goal of reducing the number of changes required to the Alby code.

The points that are touched outside of the galoy connector are:
- Define an account level field `currencyPrecision` that goes together with `currency` and represents the number of decimals for values, it defaults as 0 for sats and is set to 2 by the galoy connector
- Add a configurable step field to DuelCurrencyField to allow floating point inputs
- Use getFormattedInCurrency instead of getFormattedSats to display the values 
- Disable fiat conversion when currency != BTC, it is not clear to me if fiat2fiat conversion is wanted or worth to implement
- Change "Amount in satoshi"  with "Amount ($CURRENCY)" when currency != BTC

I've forced pushed the refactored commit to the pr branch: https://github.com/getAlby/lightning-browser-extension/commit/95291ac43b3712e80b52fb19e4e5aacd2b4c6155

openoms commented:
The history, balance and the invoice is now correctly denominated in USD, all working well. The Stablesats implementation is complete from my point view.

reneaaron reviewed (commented):
Thanks very much for this PR! 🙌

While providing a first-class USD experience for users is a great goal to strive for, it poses quite some problems that might not seem obvious in the first place and add a lot of complexity: 

The concept of sats is deeply embedded in the wallet and outside APIs (such as WebLN). Changing e.g. the receive screen to use USD as the primary currency will only work for when a user creates an invoice manually, however it will break contracts with WebLN as these APIs only knows sats and no fiat currencies. (see webln.makeInvoice() for instance)  

I wonder how we should deal with this problem 🤔 Possibly you could also add an API on Galoy side that allows to create USD invoices by using sats as the unit? (and convert it internally using the current exchange rate)

FYI: We are about to improve the invoice creation for fiat currencies in #2511 which would  improve the experience for stablesats accounts as users would be able to enter USD amounts, too. (the converted sats amount could be submitted to the API then)

For now I would prefer to keep the changes related to different UI components (ReceiveInvoice, TransactionTable, TransactionModal, etc) to an absolute minimum. I assume users will know they connected their USD denominated wallet and I am afraid changing primary units all across the application will add quite a lot of complexity when applied to all affected parts (e.g. budgets,  etc)

riccardobl commented on src/extension/background-script/connectors/galoy.ts:
```diff
@@ -419,8 +449,13 @@ class Galoy implements Connector {
         }
-        if (wallet.walletCurrency === "USD") {
-          throw new Error("USD currency support is not yet implemented.");
+        if (wallet.walletCurrency !== (this.config.currency || "BTC")) {
```
targetWallet.walletCurrency is from blink apis, currently blink has one wallet id per currency, so it should always match since the connector is initialized with the right wallet id, however i think this check is worth keeping because it will make eventual changes to the blink api break the connector instead of failing silently with unexpected results.

riccardobl commented:
> I wonder how we should deal with this problem 🤔 Possibly you could also add an API on Galoy side that allows to create USD invoices by using sats as the unit? (and convert it internally using the current exchange rate)

I think the vast majority of use cases are fine with doing the sats -> usd conversion in the galoy connector before creating the invoice, like i did in the first commit (then removed since it wasn't needed after changing the receive page with usd based input), i can revert back to this behavior and then https://github.com/getAlby/lightning-browser-extension/issues/2511 will provide the usd input as asked by @openoms 

> For now I would prefer to keep the changes related to different UI components (ReceiveInvoice, TransactionTable, TransactionModal, etc) to an absolute minimum

Since there is already the concept of currencies for balances, would you be fine with keeping the change only for transactions? If the connector uses sats internally, it should be just a visual change. 
If having `transaction.totalAmount` being USD denominated is a problem or introduces inconsistencies, i can add a new field "transaction.displayAmount" that, if set, overrides the totalAmount graphically in the UI so that connectors can display whatever they want in the transaction amount, while keeping it sat denominated internally.

Would one of these two options quality as a valid solution?

openoms commented:
Thank you for the review @reneaaron!

Having the ability to create invoices priced in USD like discussed in https://github.com/getAlby/lightning-browser-extension/issues/2511 could work as the main problem is the different accuracy of sats vs cents resulting in rounding on conversion and inconsistency in the amounts (noted in my first comment). https://github.com/getAlby/lightning-browser-extension/issues/2511 (with the USD invoice creation) would become the dependency of this PR.

The fixed USDcent denomination would need to be kept for the transaction history, would be happy with either of solutions offered by @riccardobl as the Alby maintainers see fit.

reneaaron reviewed (changes requested):
Just did another round of testing & reviews, great progress and clean implementation so far! :ok_hand: 

1) The send page currently doesn't work as it's checked against the USD balance: :thinking: 
 
![image](https://github.com/getAlby/lightning-browser-extension/assets/100827540/8d76cb00-440f-4fb5-95b4-d381802e9bea)

2) webln.makeInvoice currently leads to quite strange behavior due to different rates being used for the calculation

makeInvoice(210000) led to an invoice of 209742 sats. Do you know if there is a way to use the galoy exchange rates or denominate an invoice in sats for the USD account? 

You can use this page for testing: 
https://webln.twentyuno.net/invoice

riccardobl commented on src/extension/background-script/connectors/galoy.ts:
```diff
@@ -453,10 +497,26 @@ class Galoy implements Connector {
+    let amountSats = Number(args.amount);
+    if (currency !== "BTC") {
+      const rate = await getCurrencyRateWithCache(CURRENCIES[currency]);
+      amountSats = this.toFiatInt(amountSats * rate, currency);
```
This is actually transforming a float $ value to an int cents value. So maybe something like this would be better:
`toFiatCents`
`fromFiatCents`

riccardobl commented:
> 1. The send page currently doesn't work as it's checked against the USD balance: 🤔

this seems like a bug in alby not in this pr necessarily

riccardobl commented:
> makeInvoice(210000) led to an invoice of 209742 sats. Do you know if there is a way to use the galoy exchange rates or denominate an invoice in sats for the USD account?

I think it is fixed in the last commit, i've changed the code to use lnUsdInvoiceBtcDenominatedCreateOnBehalfOfRecipient that is either a new api or i didn't noticed it before.

riccardobl commented on src/extension/background-script/connectors/galoy.ts:
```diff
@@ -20,12 +21,15 @@ import Connector, {
   headers?: Headers; // optional for backward compatibility
   apiCompatibilityMode?: boolean; // optional for backward compatibility
   accessToken?: string; // only present in old connectors
+  currency?: GaloyCurrencies; // default is BTC
```
Done

riccardobl commented on src/extension/background-script/connectors/galoy.ts:
```diff
@@ -453,10 +497,26 @@ class Galoy implements Connector {
-        mutation lnInvoiceCreate($input: LnInvoiceCreateInput!) {
-          lnInvoiceCreate(input: $input) {
+        mutation ${mutationName}($input: ${inputTypeName}!) {
+          ${fun}(input: $input) {
```
:+1:

reneaaron reviewed (approved):
tACK

reneaaron commented:
I think this is a great start! :100: Thanks everyone involved! We're now prioritizing #2511 to improve the integration further. 

@riccardobl You have done a great job in here, thanks for your detailed analysis and thought-through suggestions on GitHub & Discord. Really appreciate your structured way of working! :muscle: If you are up for it, we'd love to get your help on #2511 as well.

riccardobl commented:
> I think this is a great start! 💯 Thanks everyone involved! We're now prioritizing #2511 to improve the integration further.
> 
> @riccardobl You have done a great job in here, thanks for your detailed analysis and thought-through suggestions on GitHub & Discord. Really appreciate your structured way of working! 💪 If you are up for it, we'd love to get your help on #2511 as well.

Thank you, it was a pleasure to contribute to the project. 
I'll start looking into #2511 and see if i can open a draft for a possible implementation

openoms commented:
Tested in the main branch now and the changes work as expected, ticking our bounty requirements which now can be considered complete: https://github.com/GaloyMoney/galoy/discussions/3950#discussioncomment-8830912

Thank you very much for the work @riccardobl and the reviews @reneaaron!

Looking forward to working together further on maintaining and improving this integration as already proposed in https://github.com/getAlby/lightning-browser-extension/issues/2511
